src/libsystemd/sd-device/device-private.c

```diff
@@ -319,7 +319,7 @@ static int device_amend(sd_device *device, const char *key, const char *value) {
                 for (const char *p = value;;) {
                         _cleanup_free_ char *word = NULL;
 
-                        r = extract_first_word(&p, &word, NULL, 0);
+                        r = extract_first_word(&p, &word, NULL, EXTRACT_RETAIN_ESCAPE);
```

Comment:
I don't think this is the right approach. If we retain the escapes in the string here, than we would have to make sure that the escapes are removed whereever the string is actually used. In particular, getting a string with escapes from `sd_device_get_devlink_next` would be totally unexpected. Instead of playing games like this, I think the only way to structure the handling of strings is to do full verification and unescaping when originally the data enters the program, and to always operate on "real" values internally, and finally always escape special values when exporting data, as appropriate for the given export format. Thus, when "exporting" the data through `sd_device_get_devlink_next`, no escaping needs to be done. When "exporting" the data through the stringified DEVLINK property, re-escape special characters.

Reply:
Besides how we should handle the string, 87a4d416e5126b6fb2528ae192a6a6a8033539ce breaks backward compatibility. Previously, `sd_device_get_devlink_first/next()` provided escaped strings. But now they does not. 

udev rules set escaped string, and sd-device itself did nothing to the input strings.
For backward compatibility, we need to keep the behavior.

Reply:
Thankfully sd_device_get_devlink_first is strictly undocumented. I guess it'll be up to the person writing the docs whenever that happens to deal with that ;)